### PR TITLE
[core][state] Record file offsets instead of logging magic token to track task log (#35572)

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
We have seen regression from `1_1_actor_calls_concurrent` and `1_1_actor_calls_async` due to the additional overheads of writing magic token to worker files when starting and finishing execution of a task. 
It's extremely bad for concurrent actors due to the contention on the same worker file. 

This PR: 
- Skips recording the task start/end for concurrent actor tasks. And augment the task log querying behaviour to be: 
  - Normal task log: we will still be able record the offsets and stream task log from dashboard + CLI
  - Normal actor task log: same as above ^
  - Async actor task + threaded actor task (max_concurrency>1) :
     - On CLI: we will raise error and let users know to use ray logs actor --id <id> for actor log.

- Revert to the old way of recording file offsets rather than writing magic tokens so that the overhead is actually lower and prevents the pollution of actor logs. 

Microbenchmark results:
| Version | 1_1_actor_calls_sync | 1_1_actor_calls_async | 1_1_actor_calls_concurrent |
|---|---|---|---|
| 2.4 Release | 2527.69 | 8374.56 | 5256.40 |
| This PR | 2404.83 | 7875.52 | 5134.41 |
| Master | 2359.93 | 6542.37 | 2760.99 |

The ~8% regression on `1_1_actor_calls_async`: I would argue for tolerating this regression since
1. The 1_n and n_n microbenchmarks of actor_calls_async don't show regression at all. 
2. This would be less of a problem if the actual actor task is not a no-op. 

In a follow-up, I will delete the dead code.https://github.com/ray-project/ray/issues/35599

## Related issue number
Closes https://github.com/ray-project/ray/issues/35598

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by zhe-thoughts** [comment]
Thanks @rickyyx . Is there a master PR that you are cherry picking from?

**Review by zhe-thoughts — APPROVED** [review]
This fixes a release blocker. Let's pick

**Comment by rickyyx** [comment]
Running some release tests for signals: https://buildkite.com/ray-project/release-tests-pr/builds/40166

**Comment by ArturNiederfahrenhorst** [comment]
@rickyyx The kickoff failed

**Comment by rickyyx** [comment]
> @rickyyx The kickoff failed

Let me try again - looks like wheel job has some issues.

**Comment by rickyyx** [comment]
Results are expected: 

```
1:1 actor calls sync per second 2591.21 +- 36.63
1:1 actor calls async per second 7563.43 +- 116.16
1:1 actor calls concurrent per second 5321.16 +- 183.82
```

**Comment by rickyyx** [comment]
@ArturNiederfahrenhorst @can-anyscale  - we should be able to merge this. PR CI is green as well.

**Comment by rickyyx** [comment]
> @ArturNiederfahrenhorst @can-anyscale - we should be able to merge this. PR CI is green as well. (microbenchmark was run but I thin there was some issue in the shutdown, and is already fixed in the branch)
